{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bICrNrRNHzjD"
      },
      "source": [
        "\n",
        "\n",
        "<img src=\"https://raw.githubusercontent.com/mahvin92/DNAcrypt-AI/main/assets/DNAcrypt%20logo/DNAcrypt-logo-3.png\" height=\"200\" align=\"center\" style=\"height:340px\">\n",
        "\n",
        "\n",
        "# DNAcrypt-AI\n",
        "**Genome-based cryptography using artificial intelligence.**\n",
        "\n",
        "\n",
        "DNAcrypt-AI is an artificial intelligence system that uses the human genome as a high-entropy foundation for generating, encrypting, and decrypting passwords and cryptographic keys. It generates and encrypts a password or key by extending a kmer-based strategy to encode alphanumeric symbols and randomly building a genome vocabulary by sampling DNA coordinates from multiple human genome assemblies. It decrypts the information by reconstructing the corresponding DNA sequences using [FAS2rDNA](https://fas2rdna.chordexbio.com/). These sequences form a complex corpus used by [Covary](https://covary.chordexbio.com/) to extract deep sequence patterns through machine learning. The resulting vector embeddings are mapped back into the unique alphanumeric and symbolic encodings from the kmer dictionary; these, in turn, represent the type and sequence of characters encrypted in the genome vocabulary, forming the original password or key.\n",
        "\n",
        "\n",
        "Version: v.0.1 (Beta, no longer supported)\n",
        "\n",
        "\n",
        "Learn more by visiting our [website](https://dnacryptai.chordexbio.com) or [GitHub repo](https://github.com/mahvin92/DNAcrypt-AI)\n",
        "\n",
        "\n",
        "This Colab version of DNAcrypt-AI was built through the [assemBold program by ChordexBio](https://assembold.chordexbio.com).\n",
        "\n",
        "---\n",
        "Versions:\n",
        "\n",
        "[v.1.1](https://colab.research.google.com/drive/1GZfnlJke6Lql41d215hSnfn4-17KB6cQ) - improved version (data builder issue fixed)\n",
        "\n",
        "[v.1.0](https://colab.research.google.com/github/mahvin92/DNAcrypt-AI/blob/main/ipynb/DNAcrypt-AI_v.1.0.ipynb) - foundational release (not supported anymore)\n",
        "\n",
        "[v.0.1](https://colab.research.google.com/github/mahvin92/DNAcrypt-AI/blob/main/ipynb/DNAcrypt-AI_beta(v.0.1).ipynb) - beta version (not supported anymore)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NSc1E0Hia7zc",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "# @title ## User configuration\n",
        "\n",
        "import json\n",
        "import os\n",
        "from google.colab import files\n",
        "\n",
        "INPUT_DIR = \"/content/DNAcrypt\" # please do not change\n",
        "OUTPUT_DIR = \"/content/DNAcrypt/outputs\" # please do not change\n",
        "project_name = \"DNAcrypt project\" # @param {\"type\":\"string\",\"placeholder\":\"e.g., My experiment\"}\n",
        "char_count = 12 # @param {\"type\":\"number\",\"placeholder\":\"12\"}\n",
        "# @markdown ####Use cases:\n",
        "Password = True # @param {type:\"boolean\"}\n",
        "Encryption = False # @param {type:\"boolean\"}\n",
        "Decryption = False # @param {type:\"boolean\"}\n",
        "input_type = 1 if Encryption else 0\n",
        "# @markdown ####Kmer dictionary:\n",
        "Default = True # @param {type:\"boolean\"}\n",
        "Custom = False # @param {type:\"boolean\"}\n",
        "kmer_type = 1 if Custom else 0\n",
        "\n",
        "# @markdown ##### For decryption, upload your files below üëá after a runtime is initiated.\n",
        "\n",
        "if Decryption:\n",
        "  input_type = 2\n",
        "else:\n",
        "  if Password == Encryption:\n",
        "    if Password:\n",
        "        print(\"‚ùå Error: Both 'Password' and 'Encryption' are selected. Please pick only one.\")\n",
        "    else:\n",
        "        print(\"‚ùå Error: Neither 'Password' nor 'Encryption' is selected. Please pick one.\")\n",
        "    raise ValueError(\"Invalid use case selection\")\n",
        "\n",
        "  # Validity checks:\n",
        "  if not (6 <= char_count <= 90):\n",
        "    print(f\"‚ùå Error: char_count ({char_count}) must be between 6 and 90.\")\n",
        "    # Stop execution in Colab cell\n",
        "    raise ValueError(\"Invalid char_count\")\n",
        "\n",
        "  if Default == Custom:\n",
        "    if Default:\n",
        "        print(\"‚ùå Error: Both 'Default' and 'Custom' are selected. Please pick only one.\")\n",
        "    else:\n",
        "        print(\"‚ùå Error: Neither 'Default' nor 'Custom' is selected. Please pick one.\")\n",
        "    raise ValueError(\"Invalid Kmer dictionary selection\")\n",
        "\n",
        "os.makedirs(INPUT_DIR, exist_ok=True)\n",
        "os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
        "%cd {OUTPUT_DIR}\n",
        "\n",
        "input_length = char_count\n",
        "\n",
        "data = {\n",
        "    \"project_name\": project_name,\n",
        "    \"input_length\": input_length,\n",
        "    \"input_type\": input_type,\n",
        "    \"kmer_type\": kmer_type\n",
        "}\n",
        "\n",
        "file_name = \"config.json\"\n",
        "\n",
        "with open(file_name, 'w') as json_file:\n",
        "    json.dump(data, json_file, indent=4)\n",
        "\n",
        "\n",
        "\n",
        "#Decryption\n",
        "if Decryption:\n",
        "  !rm -rf /content/DNAcrypt/outputs/kmer_dict.json\n",
        "  !rm -rf /content/DNAcrypt/userdata_input.txt\n",
        "  !rm -rf /content/DNAcrypt/source_dict.json\n",
        "  %cd /content/DNAcrypt\n",
        "\n",
        "  print(\"Proceed with decrypting...\")\n",
        "\n",
        "  def is_valid_metadata(obj):\n",
        "    if not isinstance(obj, dict):\n",
        "        return False\n",
        "\n",
        "    required_top = {\"data\", \"userinput\", \"usecase\"}\n",
        "    if not required_top.issubset(obj):\n",
        "        return False\n",
        "\n",
        "    # data[]\n",
        "    if not isinstance(obj[\"data\"], list) or not obj[\"data\"]:\n",
        "        return False\n",
        "    d = obj[\"data\"][0]\n",
        "    if not {\"id\", \"sequence\", \"datetime\"}.issubset(d):\n",
        "        return False\n",
        "\n",
        "    # userinput[]\n",
        "    if not isinstance(obj[\"userinput\"], list) or not obj[\"userinput\"]:\n",
        "        return False\n",
        "    u = obj[\"userinput\"][0]\n",
        "    if not {\"sample_id\", \"seq_id\", \"seq_loc\", \"description\"}.issubset(u):\n",
        "        return False\n",
        "\n",
        "    # usecase[]\n",
        "    if not isinstance(obj[\"usecase\"], list) or not obj[\"usecase\"]:\n",
        "        return False\n",
        "    if \"type\" not in obj[\"usecase\"][0]:\n",
        "        return False\n",
        "    if obj[\"usecase\"][0][\"type\"] not in (0, 1):\n",
        "        return False\n",
        "\n",
        "    return True\n",
        "\n",
        "  def is_valid_kmer_dict(obj):\n",
        "    if not isinstance(obj, dict):\n",
        "        return False\n",
        "\n",
        "    vocab = obj.get(\"kmer_vocab\")\n",
        "    if not isinstance(vocab, dict):\n",
        "        return False\n",
        "\n",
        "    for k, v in vocab.items():\n",
        "        if isinstance(k, str) and isinstance(v, list):\n",
        "            if all(isinstance(x, str) and len(x) == 5 for x in v):\n",
        "                return True\n",
        "\n",
        "    return False\n",
        "\n",
        "\n",
        "  # Upload files here\n",
        "  uploaded = files.upload()\n",
        "\n",
        "  metadata_file = None\n",
        "  kmer_dict_file = None\n",
        "\n",
        "  for fname, raw in uploaded.items():\n",
        "    if not fname.lower().endswith(\".json\"):\n",
        "        print(f\"‚ö†Ô∏è Skipped {fname} not a valid JSON file\")\n",
        "        continue\n",
        "\n",
        "    try:\n",
        "        obj = json.loads(raw.decode(\"utf-8\"))\n",
        "\n",
        "        if is_valid_metadata(obj):\n",
        "            target_name = \"DNAcrypt_metadata.json\"\n",
        "\n",
        "            if fname != target_name:\n",
        "                with open(target_name, \"w\") as f:\n",
        "                    json.dump(obj, f, indent=2)\n",
        "                metadata_file = target_name\n",
        "            else:\n",
        "                metadata_file = fname\n",
        "\n",
        "            # Reconstitue DNAcrypt-required files\n",
        "            metadata_path = \"/content/DNAcrypt/DNAcrypt_metadata.json\"\n",
        "            source_json_out = \"/content/DNAcrypt/source_dict.json\"\n",
        "            userdata_txt_out = \"/content/DNAcrypt/userdata_input.txt\"\n",
        "\n",
        "            with open(metadata_path, \"r\") as f:\n",
        "                metadata = json.load(f)\n",
        "\n",
        "            source_dict = {\n",
        "                \"data\": metadata.get(\"data\", [])\n",
        "            }\n",
        "\n",
        "            with open(source_json_out, \"w\") as f:\n",
        "                json.dump(source_dict, f, indent=2)\n",
        "\n",
        "            userinput = metadata.get(\"userinput\", [])\n",
        "\n",
        "            if userinput:\n",
        "                headers = list(userinput[0].keys())\n",
        "\n",
        "                with open(userdata_txt_out, \"w\") as f:\n",
        "                    f.write(\"\\t\".join(headers) + \"\\n\")\n",
        "                    for row in userinput:\n",
        "                        f.write(\"\\t\".join(row.get(h, \"\") for h in headers) + \"\\n\")\n",
        "\n",
        "\n",
        "        if is_valid_kmer_dict(obj):\n",
        "            target_name = \"kmer_dict.json\"\n",
        "\n",
        "            if fname != target_name:\n",
        "                with open(target_name, \"w\") as f:\n",
        "                    json.dump(obj, f, indent=2)\n",
        "                kmer_dict_file = target_name\n",
        "\n",
        "            else:\n",
        "                kmer_dict_file = fname\n",
        "\n",
        "            !mv /content/DNAcrypt/kmer_dict.json /content/DNAcrypt/outputs/kmer_dict.json\n",
        "\n",
        "        else:\n",
        "\n",
        "          # Determine correct static kmer dictionary to use\n",
        "          metadata_path = \"/content/DNAcrypt/DNAcrypt_metadata.json\"\n",
        "\n",
        "          if not os.path.exists(metadata_path):\n",
        "              raise FileNotFoundError(\"‚ùå DNAcrypt_metadata.json not found\")\n",
        "\n",
        "          with open(metadata_path, \"r\") as f:\n",
        "              metadata = json.load(f)\n",
        "\n",
        "          usecase = metadata.get(\"usecase\", [])\n",
        "          if not usecase or \"type\" not in usecase[0]:\n",
        "              raise ValueError(\"‚ùå You have a corrupted data\")\n",
        "\n",
        "          usecase_type = usecase[0][\"type\"] # use to rebuild dictionary after json folder is fetched\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"‚ùå Failed to read {fname}: {e}\")\n",
        "\n",
        "\n",
        "print(f\"Proceeding to install DNAcrypt-AI. Please wait...\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fwN7Fj18X3sG",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "# @title ## Install DNAcrypt-AI\n",
        "%%capture\n",
        "\n",
        "%cd /content/DNAcrypt/\n",
        "\n",
        "install_dnacrypt = \"https://github.com/mahvin92/DNAcrypt-AI.git\"\n",
        "!git init\n",
        "!git remote add origin {install_dnacrypt}\n",
        "!git config core.sparseCheckout true\n",
        "!echo \"json/*\" >> .git/info/sparse-checkout\n",
        "!echo \"source/*\" >> .git/info/sparse-checkout\n",
        "!git pull origin main\n",
        "!mv source/* . 2>/dev/null\n",
        "!rm -rf source .git\n",
        "!cp /content/DNAcrypt/json/assembly_dict.json /content/DNAcrypt/outputs\n",
        "\n",
        "# Get dependencies\n",
        "!pip install /content/DNAcrypt/install/codeenigma_runtime-1.2.0-py3-none-any.whl\n",
        "!pip install pandas pyfaidx tqdm\n",
        "!apt-get update -qq\n",
        "!apt-get install -y samtools\n",
        "import sys, builtins\n",
        "import os\n",
        "import sys\n",
        "import re\n",
        "import math\n",
        "from itertools import product\n",
        "!pip install umap-learn\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "from sklearn.preprocessing import StandardScaler, LabelEncoder\n",
        "from sklearn.decomposition import PCA\n",
        "from sklearn.manifold import TSNE\n",
        "import umap\n",
        "from tensorflow.keras import layers, models\n",
        "from scipy.cluster.hierarchy import linkage, dendrogram\n",
        "import seaborn as sns\n",
        "from scipy.spatial.distance import pdist, squareform\n",
        "import zipfile\n",
        "from datetime import datetime\n",
        "import sys, builtins"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wwj48rW8bDaK",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "# @title ##Workflow detection\n",
        "\n",
        "import json\n",
        "from pathlib import Path\n",
        "import os\n",
        "from google.colab import files\n",
        "\n",
        "config_path = Path('/content/DNAcrypt/outputs/config.json')\n",
        "\n",
        "# Encryption\n",
        "if not Decryption:\n",
        "  sys.path.append(\"/content/DNAcrypt\")\n",
        "  builtins.exit = sys.exit\n",
        "  import dict_router\n",
        "\n",
        "else:\n",
        "  if usecase_type == 0:\n",
        "    !cp /content/DNAcrypt/json/kmer_dict_key.json /content/DNAcrypt/outputs/kmer_dict.json\n",
        "\n",
        "  elif usecase_type == 1:\n",
        "    !cp /content/DNAcrypt/json/kmer_dict_enc.json /content/DNAcrypt/outputs/kmer_dict.json\n",
        "\n",
        "  else:\n",
        "    raise ValueError(\"‚ùå Invalid encryption data\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-SgUktVLnXLS",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "# @title ## Genome encryption\n",
        "\n",
        "if not Decryption:\n",
        "  import json\n",
        "  import sys\n",
        "  from pathlib import Path\n",
        "\n",
        "  sys.path.append(\"/content/DNAcrypt\")\n",
        "  builtins.exit = sys.exit\n",
        "  import num_crypt\n",
        "  import genome_keyring"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title ## Sequence reconstruction (FAS2rDNA)\n",
        "\n",
        "INPUT_DIR = \"/content/fas2rdna\" # please do not change\n",
        "OUTPUT_DIR = \"/content/fas2rdna/outputs\" # please do not change\n",
        "os.makedirs(INPUT_DIR, exist_ok=True)\n",
        "os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
        "!cp /content/DNAcrypt/userdata_input.txt /content/fas2rdna\n",
        "%cd {INPUT_DIR}\n",
        "\n",
        "# Initiate FAS2rDNA\n",
        "sys.path.append(\"/content/DNAcrypt\")\n",
        "builtins.exit = sys.exit\n",
        "import fas2rdna\n",
        "\n",
        "# Append anchor\n",
        "import append_anchor"
      ],
      "metadata": {
        "id": "Xnh9_uOSPddW",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cTKQf4dMcHzk",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "# @title ## Machine learning (Covary)\n",
        "%cd /content/\n",
        "\n",
        "# Clone Covary-encoder\n",
        "COVARY_DIR = \"/content/Covary-encoder\" # please do not change\n",
        "os.makedirs(COVARY_DIR, exist_ok=True)\n",
        "\n",
        "%cd Covary-encoder\n",
        "!cp /content/fas2rdna/outputs/userdata_input.fasta /content/Covary-encoder/\n",
        "os.rename(\"/content/Covary-encoder/userdata_input.fasta\", \"/content/Covary-encoder/input_seq.fasta\")\n",
        "\n",
        "# Pre-processing for Covary inference\n",
        "sys.path.append(\"/content/DNAcrypt\")\n",
        "builtins.exit = sys.exit\n",
        "import clean_seq\n",
        "\n",
        "# Genetic encoding with Covary_encoder\n",
        "sys.path.append(\"/content/DNAcrypt\")\n",
        "builtins.exit = sys.exit\n",
        "import covary_encoder\n",
        "\n",
        "# Machine learning with Covary-ML\n",
        "print(f\"Deconvoluting your data now on Covary. Please wait...\")\n",
        "sys.path.append(\"/content/DNAcrypt\")\n",
        "builtins.exit = sys.exit\n",
        "import covary_ml"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title ## Data builder\n",
        "\n",
        "import os\n",
        "import shutil\n",
        "import json\n",
        "from google.colab import files\n",
        "\n",
        "# Prompt decrypted key\n",
        "sys.path.append(\"/content/DNAcrypt\")\n",
        "builtins.exit = sys.exit\n",
        "import dnacrypt_builder\n",
        "\n",
        "if Decryption:\n",
        "  print(\"Decryption finished ...\")\n",
        "else:\n",
        "  sys.path.append(\"/content/DNAcrypt\")\n",
        "  builtins.exit = sys.exit\n",
        "  import dnacrypt_compiler"
      ],
      "metadata": {
        "id": "t2PO_13dVan4",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PrXYXyraFtF-"
      },
      "source": [
        "# Instructions\n",
        "\n",
        "## Quick notice:\n",
        "DNAcrypt-AI generates a password (alphanumeric + symbol) or cryptographic key (alphanumeric) randomly, with a user-defined length. The password or key is then encrypted using the coordinates of randomly sampled length-variable DNA sequences found in the human genome (referenced to assemblies hg19 and hg38). DNAcrypt-AI decrypts the data using a high-throughput DNA sequence reconstitution pipeline ([FAS2rDNA](https://github.com/mahvin92/FAS2rDNA)) and sequence-informed machine learning model ([Covary](https://github.com/mahvin92/Covary)). It does not support custom password generation, however, it supports high entropy random sequence assignment.\n",
        "\n",
        "DNAcryt-AI supports alphanumeric encodings ```a -> z, A -> Z, 0 -> 9``` (for both Password and Encryption) and symbolic encodings such as ```!@#$%^&*()-_+=‚Äù``` (for Password only).\n",
        "\n",
        "## Supported human genome assemblies:\n",
        "The hg19 and hg38 genome assemblies are used as reference to encrypt and decrypt your password/key. Therefore, DNAcrypt-AI uses a biological and native genome sequences to store and read your secret information. Although current works is already being undertaken to expand this support to multi-species genome assemblies to increase the genome vocabularies of DNAcrypt-AI. Reach out to us if you have any recommendations or suggestions to improve this project.\n",
        "\n",
        "## Generating and encrypting a password/key:\n",
        "DNAcrypt-AI is very intuitive to use. To generate and encrypt a password or a key:\n",
        "1. Create a user configuration:\n",
        "- char_count: string length\n",
        "- Use cases: *Password* = alphanumeric + symbol; *Encryption* = alphanumeric only\n",
        "2. Run DNAcrypt-AI: ```Runtime -> Run all```\n",
        "3. Store your encrypted data (DNAcrypt_metadata.json and/or kmer_dict.json if you used to custom kmer dictionary)\n",
        "- These are automatically downloaded. However, if your browser is preventing the download, you can get them directly from ```/content/``` and/or ```/content/DNAcrypt/outputs/``` through the File browser.\n",
        "\n",
        "## Decrypting a password/key:\n",
        "Decrypting information is very easy. Follow the steps below:\n",
        "1. Modify the user configuration\n",
        "- Use cases: select *Decryption*\n",
        "2. Run DNAcrypt-AI: ```Runtime -> Run all```\n",
        "3. Upload your encrypted data\n",
        "- Note that in addition to the original 'DNAcrypt_metadata.json', you are required to upload the 'kmer_dict.json' if you encrypted your data using a customized kmer dictionary.\n",
        "4. Wait for the key to be decrypted\n",
        "- This may take some time, roughly about 15 minutes or less, depending on the genome vocabularies you have.\n",
        "\n",
        "## Encrypting and decrypting using a custom kmer dictionary:\n",
        "a. Encryption\n",
        "1. In addition to char_count and a use case, create a user configuration by selecting *Custom* under Kmer dictionary.\n",
        "2. Run DNAcrypt-AI as usal\n",
        "3. Store both the *DNAcrypt_metadata.json* and *kmer_dict.json* files for decryption later.\n",
        "\n",
        "b. Decryption\n",
        "1. Select *Decryption* as use case in the user configuration\n",
        "2. Run DNAcrypt-AI as usal\n",
        "3. Upload the *DNAcrypt_metadata.json* and *kmer_dict.json* files\n",
        "\n",
        "## Encryption data:\n",
        "Your encrpytion data MUST NOT be modified. Any loss, modification, or unnecessary addition will affect the stored information. Tampering with the encrypted data (both *DNAcrypt_metadata.json* and *kmer_dict.json*) will affect the recovery of your password/key. However, you can rename the files as you wish.\n",
        "\n",
        "The encrypted data is light and does not take too much of your storage (maximum of 200kb). You can store your encryption as follows:\n",
        "1. Printed copy - private and secure but requires re-encoding to a digital copy later before using\n",
        "2. Digital copy - ready to use but also allows others to decrypt once they get a copy of your file/s\n",
        "\n",
        "## Training DNAcrypt-AI:\n",
        "DNAcrypt-AI (this notebook version) runs entirely in your browser; the developer does not collect nor store your data in the cloud or local server. Furthermore, DNAcrypt-AI does not index your encrypted data, nor does it utilize a pre-packed knowledge base or vector store of genomic vocabularies (such as encoded human genome sequences or k-mer dictionaries used in alphanumeric/symbol encoding). By design through [Covary](https://github.com/mahvin92/Covary), users train and run DNAcrypt-AI locally, ensuring their data remains exclusively theirs. Please note that machine learning inference is computationally intensive, so expect a processing time of approximately 15 minutes to successfully decrypt your password or key.\n",
        "\n",
        "## Troubleshooting:\n",
        "1. Decrypted password/key does not match the expected alphanumeric-symbol sequence\n",
        "- Re-run the decryption or restart your session ```Runtime -> Restart session and run all```. If this does not fix the issue, check if you're using a GPU or CPU for training. It has been reported that decryption varies between GPU- and CPU-based inferences. Try changing you runtime from GPU to CPU and *vice versa*.\n",
        "2. DNAcrypt-AI session crashed due to RAM exhaustion\n",
        "- Machine learning is a RAM-hungry process, ensure that you are connected to a GPU support (e.g., T4)\n",
        "3. DNAcrypt-AI is taking so long in reconstucting my genome corpus\n",
        "- The reconstruction of DNA sequences is loaded to FAS2rDNA, which requires to download either or both the hg19 and hg38 human genome assemblies. These assemblies are quite massive (>3 GB each) in size. Please wait for a moment for these assemblies to be fetched and indexed\n",
        "4. DNAcrypt-AI is not downloading the required file (e.g., *DNAcrypt_metadata.json*)\n",
        "- If DNAcrypt-AI executed without an error and yet did not produce a download, it means that your browser is blocking the process. You can fetch your encrypted files from ```/content``` or ```/content/DNAcrypt/outputs/``` directory through the File browser.\n",
        "5. I obtained a low entropy sequence\n",
        "- DNAcrypt-AI estimates the entropy of the generated password (the random assignment and sequence of alphanumeric-symbol charcaters), not necessarily the strength of the encryption. If you are not satisfied with the entropy estimate, you can always run DNAcrypt-AI again until you arrive at a desirable entropy. If you encounter issues while running DNAcrypt-AI multiple times, restarting your session can fix most issues ```Runtime -> Restart session and run all```\n",
        "\n",
        "## Bug:\n",
        "\n",
        "If you encounter any bugs, please open a ticket [here](https://github.com/mahvin92/DNAcrypt-AI/issues/new) or request a support [here](https://dnacryptai.chordexbio.com).\n",
        "\n",
        "\n",
        "## Funding (unofficial):\n",
        "\n",
        "[ChordexBio](https://chordexbio.com)\n",
        "\n",
        "\n",
        "## License:\n",
        "DNAcrypt-AI uses FAS2rDNA and Covary, with permission from ChordexBio. Your usage of DNAcrypt-AI and its dependencies may be limited, [read the full license here.](https://github.com/mahvin92/DNAcrypt-AI)\n",
        "\n",
        "\n",
        "## Footnote:\n",
        "DNAcrypt-AI is powered by [ChordexBio](https://chordexbio.com), [FAS2rDNA](https://github.com/mahvin92/FAS2rDNA), [Covary](https://github.com/mahvin92/Covary), [assemBold program](https://assembold.chordexbio.com) and [CodeEnigma](https://github.com/KrishnanSG/codeenigma), made with Python, and tested using Google Colab ‚ù§Ô∏è"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
